daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala

```diff
@@ -1406,28 +1406,50 @@ private[lf] object DecodeV1 {
       BuiltinFunctionInfo(GREATER_DATE, BGreaterDate),
       BuiltinFunctionInfo(EQUAL, BEqual, minVersion = genMap),
       BuiltinFunctionInfo(EQUAL_LIST, BEqualList),
-      BuiltinFunctionInfo(EQUAL_CONTRACT_ID, BEqualContractId),
-      // FIXME https://github.com/digital-asset/daml/issues/3752
-      // Constrain max version of the following 'EQUAL_' builtin once
-      // generic equality is handled by the compiler
       BuiltinFunctionInfo(EQUAL_INT64, BEqual, implicitParameters = List(TInt64)),
       BuiltinFunctionInfo(
         EQUAL_DECIMAL,
         BEqualNumeric,
         maxVersion = Some(numeric),
         implicitParameters = List(TNat.Decimal)
       ),
-      BuiltinFunctionInfo(EQUAL_NUMERIC, BEqualNumeric, minVersion = numeric),
-      BuiltinFunctionInfo(EQUAL_TEXT, BEqual, implicitParameters = List(TText)),
-      BuiltinFunctionInfo(EQUAL_TIMESTAMP, BEqual, implicitParameters = List(TTimestamp)),
-      BuiltinFunctionInfo(EQUAL_DATE, BEqual, implicitParameters = List(TDate)),
-      BuiltinFunctionInfo(EQUAL_PARTY, BEqual, implicitParameters = List(TParty)),
-      BuiltinFunctionInfo(EQUAL_BOOL, BEqual, implicitParameters = List(TBool)),
+      BuiltinFunctionInfo(
+        EQUAL_NUMERIC,
+        BEqualNumeric,
+        minVersion = numeric,
+        maxVersion = Some(genMap)),
+      BuiltinFunctionInfo(
+        EQUAL_TEXT,
+        BEqual,
+        maxVersion = Some(genMap),
+        implicitParameters = List(TText)),
+      BuiltinFunctionInfo(
+        EQUAL_TIMESTAMP,
+        BEqual,
+        maxVersion = Some(genMap),
+        implicitParameters = List(TTimestamp)),
+      BuiltinFunctionInfo(
+        EQUAL_DATE,
+        BEqual,
+        maxVersion = Some(genMap),
+        implicitParameters = List(TDate)),
+      BuiltinFunctionInfo(
+        EQUAL_PARTY,
+        BEqual,
+        maxVersion = Some(genMap),
+        implicitParameters = List(TParty)),
+      BuiltinFunctionInfo(
+        EQUAL_BOOL,
+        BEqual,
+        maxVersion = Some(genMap),
+        implicitParameters = List(TBool)),
       BuiltinFunctionInfo(
         EQUAL_TYPE_REP,
         BEqual,
         minVersion = typeRep,
+        maxVersion = Some(genMap),
         implicitParameters = List(TTypeRep)),
+      BuiltinFunctionInfo(EQUAL_CONTRACT_ID, BEqualContractId, maxVersion = Some(genMap)),
```

Comment:
Why do we not map this to `BEqual` with an implicit parameter?

Reply:
The type of ``EQUAL_CONTRACT_ID`` contains a universal quantifier (``∀a. ContractId a -> ContractId a -> Bool``). I cannot represent it, using implicit parameters. 
It is the same for `EQUAL_NUMERIC`.

 I do not think it is worthwhile to extends the implicit parameter feature to handle this two cases.

Reply:
Ah I see, thanks for the explanation! Might be worth capturing in a comment somewhere.

Reply:
It's also the reason these two had separate representations in the Haskell AST 🙃